### PR TITLE
Move Mutation Callbacks to MutationOptions

### PR DESCRIPTION
**Pull request by mainawycliffe**
This PR moves Mutation Callbacks to under `MutationOptions` so that more logic is in the GraphQL Client instead of widget side and to be more in line with apollo, where these callbacks are directly under MutationOptions. This also attempts to expose some of those callbacks so that they can be used by developers using the graphql client (#278). 

On top of that, I have added a `onError` callback to compliment both the `update` and `onComplete` callbacks just like in apollo. Open to the idea of improvements to this PR and feedback from those who will test it.

#### Breaking Changes

This going to be a breaking change as the Mutation Callbacks needs to be called directly under MutationOptions instead of the Widget Builder.

before:

```dart
Mutation(
  options: MutationOptions(
    document: addStar, // this is the mutation string you just created
  ),
  builder: (
    RunMutation runMutation,
    QueryResult result,
  ) {
    return FloatingActionButton(
      onPressed: () => runMutation({
        'starrableId': <A_STARTABLE_REPOSITORY_ID>,
      }),
      tooltip: 'Star',
      child: Icon(Icons.star),
    );
  },
  // you can update the cache based on results
  update: (Cache cache, QueryResult result) {
    return cache;
  },
  // or do something with the result.data on completion
  onCompleted: (dynamic resultData) {
    print(resultData);
  },
);
```

after:

```dart
Mutation(
  options: MutationOptions(
    document: addStar,
     // you can update the cache based on results
    update: (Cache cache, QueryResult result) {
      return cache;
    },
    // or do something with the result.data on completion
    onCompleted: (dynamic resultData) {
      print(resultData);
    },
  ),
  builder: (
    RunMutation runMutation,
    QueryResult result,
  ) {
    return FloatingActionButton(
      onPressed: () => runMutation({
        'starrableId': <A_STARTABLE_REPOSITORY_ID>,
      }),
      tooltip: 'Star',
      child: Icon(Icons.star),
    );
  },

);
```

#### Fixes / Enhancements

- This fixes issues #365 and #278

### CODE REVIEW

File: packages/graphql/lib/src/core/query_options.dart [hunk 1]
```diff
@@ -118,12 +124,106 @@ class MutationOptions extends BaseOptions {
     FetchPolicy fetchPolicy,
     ErrorPolicy errorPolicy,
     Map<String, dynamic> context,
```
Comment (smkhalsa) [inline packages/graphql/lib/src/core/query_options.dart]:
```suggestion
    Map<String, dynamic> context,
    Object optimisticResult,
```

File: packages/graphql/lib/src/core/query_options.dart [hunk 2]
```diff
@@ -118,12 +124,106 @@ class MutationOptions extends BaseOptions {
...
+    this.onCompleted,
+    this.update,
+    this.onError,
   }) : super(
           policies: Policies(fetch: fetchPolicy, error: errorPolicy),
           document: document,
           variables: variables,
           context: context,
```
Comment (smkhalsa) [inline packages/graphql/lib/src/core/query_options.dart]:
```suggestion
          context: context,
          optimisticResult: optimisticResult,
```

File: packages/graphql/lib/src/core/query_options.dart (hunk 1 above)
Comment (mainawycliffe) [reply packages/graphql/lib/src/core/query_options.dart]: I think this needs to be more thought out implementation and should be a separate issue. It is not as easy as moving existing APIs from one package to another.

**Comment by micimize** [comment]
Rebased and added a fetchmore regression and testing fixes from #429

@HofmannZ This is good to go other than CI hanging

**Pull request merged**